Keystatic guide (#4217)
withastro/docs: src/content/docs/en/guides/cms/keystatic.mdx

      config: keystaticConfig,
    })
    
    export const prerender = false
    ```

4. To launch your Keystatic Admin UI dashboard, start Astro's dev server:

    ```bash
    npm run dev
    ```

5. Visit `http://127.0.0.1:3000/keystatic` in the browser to see the Keystatic Admin UI running.

## Creating a new post

1. In the Keystatic Admin UI dashboard, click on the “Posts” collection.

 2. Use the button to create a new post. Add the title "My First Post" and some content, then save the post.
 
 3. This post should now be visible from your "Posts" collection. You can view and edit your individual posts from this dashboard page.
 
 4. Return to view your Astro project files. You will now find a new `.mdoc` file inside the `src/content/posts` directory for this new post:

		<FileTree title="Project Structure">
		- src/
		  - content/
		    - posts/
		      - **my-first-post.mdoc**
		</FileTree>

5. Navigate to that file in your code editor and verify that you can see the Markdown content you entered. For example: 

		```markdown
		---
		title: My First Post
		---
		
		This is my very first post. I am **super** excited!
		```

## Rendering Keystatic content

> Use Astro's Content Collections API to [query and display your posts and collections](en/guides/content-collections/#querying-collections), just as you would in any Astro project.

Reviewer: ```suggestion
Use Astro's Content Collections API to [query and display your posts and collections](/en/guides/content-collections/#querying-collections), just as you would in any Astro project.
```

title: Keystatic & Astro
description: Add content to your Astro project using Keystatic as a CMS
type: cms
service: Keystatic
i18nReady: false
---
import FileTree from '~/components/FileTree.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import Aside from '~/components/Aside.astro'

[Keystatic](https://keystatic.com/) is an open source, headless content-management system that allows you to structure your content and sync it with GitHub.

:::tip
If you're starting a **new Astro + Keystatic project from scratch**, you can use the [Keystatic CLI](https://keystatic.com/docs/quick-start#keystatic-cli) to generate a new project in seconds:

```bash
npm create @keystatic@latest
```

Select the Astro template, and you'll be ready to [deploy](#deploying-keystatic--astro)!
:::

## Prerequisites

> This guide assumes you already have **an existing Astro project** in SSR running on [`output: 'hybrid'`](<>). If you don’t, you can follow the instructions in our [Installation guide](/en/install/auto/).

Author: You don't need a Keystatic account — unless you're talking about the `cloud` product that's a semi-separate thing.

It's almost ready for preview/testing, and if you're curious, we have a `cloud` channel in our [Keystatic discord](https://keystatic.com/chat) :)

But it's not relevant for this guide, no need to mention it here, at least not until it's a live thing.

:::note
If you intend to sync Keystatic's data with GitHub, you will also need **a GitHub account with `write` permissions** on the repository for this project.
:::

## Installing dependencies

You're going to need to install a few packages to get Keystatic going.

Keystatic outputs Markdoc and has a dependency on React, so first you need to add Astro's first-party integrations for both:

```bash
npx astro add react markdoc
```

You will also need two Keystatic packages:

<PackageManagerTabs>
  <Fragment slot="npm">
  ```shell
  npm install @keystatic/core @keystatic/astro
  ```
  </Fragment>
  <Fragment slot="pnpm">
  ```shell
  pnpm install @keystatic/core @keystatic/astro
  ```
  </Fragment>
  <Fragment slot="yarn">
  ```shell
  yarn add @keystatic/core @keystatic/astro
  ```
  </Fragment>
</PackageManagerTabs>

## Creating a Keystatic config file

> Keystatic needs a config file. This is where you can connect a project with a specific GitHub repository and define a content schema.

Author: ```suggestion
Keystatic needs a config file. This is where you can connect a project with a specific GitHub repository and/or define a content schema.
```

Not sure if I like the `and/or` syntax, but technically you dont' *need* to connect the project to GitHub. You need however to define some content schemas before yous can do anything with Keystatic.

> Keystatic needs a config file. This is where you can connect a project with a specific GitHub repository and define a content schema.

Reviewer: OK, then what about something like (include an extra bit at the end if you like something like that, or just stop early):

> A Keystatic config file is required to define your content schema. This file will also allow you to connect a project to a specific GitHub repository (if you decide to do so).

And fwiw, technically in the original version, we're only saying above that the CONFIG file is required in order to do these things, not that these things are both required things to do! But, separation like this can help you avoid an and/or situation.)

> Keystatic needs a config file. This is where you can connect a project with a specific GitHub repository and define a content schema.

Author: Yes this is great!

```suggestion
A Keystatic config file is required to define your content schema. This file will also allow you to connect a project to a specific GitHub repository (if you decide to do so).
```

Create a file called `keystatic.config.ts` in the root of the project:

```ts {8-25}
// keystatic.config.ts
import { config, fields, collection } from '@keystatic/core';

export default config({
  storage: {
    kind: 'local',
  },
  collections: {
    posts: collection({
      label: 'Posts',
      slugField: 'title',
      path: 'src/content/posts/*',
      format: { contentField: 'content' },
      schema: {
        title: fields.slug({ name: { label: 'Title' } }),
        content: fields.document({
          label: 'Content',
          formatting: true,
          dividers: true,
          links: true,
          images: true,
        }),
      },
    }),
  },
});
```

You've exported a config object wrapped in the `config` function imported from `@keystatic/core`.

For now, the `storage` strategy is set to `local`, and there is a single “posts” `collection`.

> 

Author: Yeah, 100% agree!

> :::note
> The schema of your Keystatic collections should match the collections you've defined in Astro's [content collections](/en/guides/content-collections/#defining-collections).
> 
> We're working on a way to automatically sync Keystatic collections with Astro's content collections... Stay tuned! 🤗
> :::

Reviewer: ```suggestion
:::note[already using content collections?]
If you are already using [content collections](/en/guides/content-collections/#defining-collections) in your Astro project, then update the schema above to exactly match the collection(s) defined in your existing schema.
:::
```

Author: Annnnd I've just learned how to pass a `title` to the `:::note` component without importing the `Aside` and rendering the JSX component 😅

> This is all Keystatic needs to start managing content, configuration-wise.
> 
> Now, you need to display the Keystatic Admin UI on our site!

Reviewer: ```suggestion
Keystatic is now configured to manage your content and enforce your schema.
```

Author: ```suggestion
Keystatic is now configured to manage your content based on your schema.
```

Reviewer: I like this! (Even if I had to read it three times at this time of night to figure out how it was different!)  😅  Ship it!

## Setting up the Keystatic Admin UI dashboard

> The Keystatic Admin UI is a React app that runs in your Astro project. The following steps will set it up to live in your `/pages/` folder, and provide a dashboard for managing your content.

Reviewer: ```suggestion
The Keystatic Admin UI is a React app that runs in your Astro project. 

The following steps will create a dashboard that lives in `src/pages/keystatic/` for managing your site content.
```

    ```ts
    // keystatic.page.ts
    import { makePage } from '@keystatic/astro/ui'
    import keystaticConfig from './keystatic.config'
    
    export const Keystatic = makePage(keystaticConfig)
    ```

2. Create a new page called `src/pages/keystatic/[...params].astro` and mount the component on the client side only:

    ```astro
    ---
    // src/pages/keystatic/[...params].astro
    import { Keystatic } from '../../../keystatic.page'
    export const prerender = false
    ---
    
    <Keystatic client:only />
    ```

3. Create a new file called `src/pages/api/keystatic/[...params].ts` that will create `API routes` for Keystatic's Admin UI.

    ```jsx
    // src/pages/api/keystatic/[...params].ts
    import { makeHandler } from '@keystatic/astro/api'
    import keystaticConfig from '../../../../keystatic.config'
    
    export const all = makeHandler({
      config: keystaticConfig,
    })
    
    export const prerender = false
    ```

4. To launch your Keystatic Admin UI dashboard, start Astro's dev server:

    ```bash
    npm run dev
    ```

5. Visit `http://127.0.0.1:3000/keystatic` in the browser to see the Keystatic Admin UI running.

## Creating a new post

> Try and visit the `/keystatic` page in the browser one more time, and click on the “Posts” collection.
> 
> Go ahead, and create a new post.
> 
> :::note
> In our Keystatic config file, you've set the `path` property for our `posts` collection to `src/content/posts/*`.
> 
> As a result, creating a new post from the Keystatic Admin UI should create a new `content` directory in the `src` directory, with the new post `.mdoc` file inside.
> :::
> 
> If everything worked correctly, you'll find your new post inside the `src/content/posts` directory:

Reviewer: ```suggestion
1. In the Keystatic Admin UI dashboard, click on the “Posts” collection.

 2. Use the button to create a new post. Add the title "My First Post" and some content, then save the post.
 
 3. This post should now be visible from your "Posts" collection. You can view and edit your individual posts from this dashboard page.
 
 4. Return to view your Astro project files. You will now find a new `.mdoc` file inside the `src/content/posts` directory for this new post:
```

Author: Much clearer like that ✨

> <FileTree title="Project Structure">
> - src/
>   - content/
>     - posts/
>       - **my-first-post.mdoc**
> </FileTree>
> 
> That Markdoc file should look something like this:
> 
> ```markdown
> ---
> title: My First Post
> ---
> 
> This is my very first post. I am **super** excited.
> ```
> 
> Niiiice ✨
> 
> Time to display that post on the frontend now.

Reviewer: ``````suggestion
		<FileTree title="Project Structure">
		- src/
		  - content/
		    - posts/
		      - **my-first-post.mdoc**
		</FileTree>

5. Navigate to that file in your code editor and verify that you can see the Markdown content you entered. For example: 

		```markdown
		---
		title: My First Post
		---
		
		This is my very first post. I am **super** excited!
		```
``````

## Rendering Keystatic content

> :::tip
> While Keystatic provides its own [Reader and Renderer APIs](https://keystatic.com/docs/reader-api) to pull data for your frontend, it also works great with Astro's [content collections](/en/guides/content-collections/)!
> :::
> 
> ### Displaying a collection list
> 
> Since you instructed Keystatic to manage `posts` in `src/content/posts`, they'll be available as Astro's content collections as the `posts` collection slug:
> 
> ```tsx {4}
> ---
> import { getCollection } from 'astro:content'
> 
> const posts = await getCollection('posts')
> ---
> 
> <main>
>   <pre>{JSON.stringify(posts, null, 2)}</pre>
> </main>
> 
> ```
> 
> ### Displaying a single collection entry
> 
> You can also get a single collection entry by its slug:
> 
> ```tsx {4}
> ---
> import { getEntry } from 'astro:content'
> 
> const post = await getEntry('posts', 'my-first-post')
> ---
> 
> <main>
>   <pre>{JSON.stringify(post, null, 2)}</pre>
> </main>
> ```
> 
> ### Rendering document field content
> 
> If you examine the output from our `post` entry above, you will notice that the `body` field — which represents the `content` of our post — is displaying raw Markdown.
> 
> But you probably want to render HTML instead on your page.
> 
> Well, good news: Astro provides a `<Content />` component that can do the heavy lifting of converting rich text data into properly formatted markup:

Reviewer: ``````suggestion
Use Astro's Content Collections API to [query and display your posts and collections](en/guides/content-collections/#querying-collections), just as you would in any Astro project.

### Displaying a collection list

The following example displays a list of each post title, with a link to an individual post page.

```tsx {4}
---
import { getCollection } from 'astro:content'

const posts = await getCollection('posts')
---
<ul>
  {posts.map(post => (
    <li>
      <a href={`/posts/${post.slug}`}>{post.data.title}</a>
    </li>
  ))}
</ul>
```

### Displaying a single entry

To display content from an individual post, you can import and use the `<Content />` component to [render your content to HTML)(/en/guides/content-collections/#rendering-content-to-html):
``````
OK... :sweat_smile: 

I really think at this point, we're back on Astro territory, and... I think we can get away with just saying, "Dude! It's content collections now!" So, see how this feels...

Author: I agree. One really neat thing about Keystatic is it can just do its thing and get out of the way. Letting Astro handle things the Astro way is the _normal_ pathway here, and treating it as the normal thing is indeed adequate 👍

```tsx {4-5}
---
import { getEntry } from 'astro:content'

const post = await getEntry('posts', 'my-first-post')
const { Content } = await post.render()
---

<main>
  <h1>{post.data.title}</h1>
  <Content />
</main>

```

> 

Reviewer: ```suggestion

For more information on querying, filtering, displaying your collections content and more, see the full content [collections documentation](/en/guides/content-collections/).
```
Again, given that this is now just content collections, let's make an easy off-ramp to doing more things with your content!

title: Keystatic & Astro
description: Add content to your Astro project using Keystatic as a CMS
type: cms
service: Keystatic
i18nReady: false
---
import FileTree from '~/components/FileTree.astro'
import PackageManagerTabs from '~/components/tabs/PackageManagerTabs.astro'
import Aside from '~/components/Aside.astro'

[Keystatic](https://keystatic.com/) is an open source, headless content-management system that allows you to structure your content and sync it with GitHub.

:::tip
If you're starting a **new Astro + Keystatic project from scratch**, you can use the [Keystatic CLI](https://keystatic.com/docs/quick-start#keystatic-cli) to generate a new project in seconds:

```bash
npm create @keystatic@latest
```

> Select the Astro template, and you'll be ready to go!

Reviewer 2: ```suggestion
Select the Astro template, and you'll be ready to [deploy](#deploying-keystatic--astro)!
```

```tsx {4}
---
import { getCollection } from 'astro:content'

const posts = await getCollection('posts')
---

<main>
  <pre>{JSON.stringify(posts, null, 2)}</pre>
</main>

```

### Displaying a single collection entry

You can also get a single collection entry by its slug:

```tsx {4}
---
import { getEntry } from 'astro:content'

const post = await getEntry('posts', 'my-first-post')
---

<main>
  <pre>{JSON.stringify(post, null, 2)}</pre>
</main>
```

### Rendering document field content

If you examine the output from our `post` entry above, you will notice that the `body` field — which represents the `content` of our post — is displaying raw Markdown.

But you probably want to render HTML instead on your page.

Well, good news: Astro provides a `<Content />` component that can do the heavy lifting of converting rich text data into properly formatted markup:

```tsx {4-5}
---
import { getEntry } from 'astro:content'

const post = await getEntry('posts', 'my-first-post')
const { Content } = await post.render()
---

<main>
  <h1>{post.data.title}</h1>
  <Content />
</main>

```

## Deploying Keystatic + Astro

> Because Keystatic needs to run serverside code and use Node.js APIs, you'll need to add an [Astro adapter](/en/guides/server-side-rendering/#adding-an-adapter) to deploy your project.

Reviewer 2: We added this to prerequisites

```suggestion
To deploy your website, visit our [deployment guides](/en/guides/deploy/) and follow the instructions for your preferred hosting provider.
```

We're working on a way to automatically sync Keystatic collections with Astro's content collections... Stay tuned! 🤗
:::

This is all Keystatic needs to start managing content, configuration-wise.

Now, you need to display the Keystatic Admin UI on our site!

## Keystatic Admin UI pages

In our `pages` directory, every route within the `/keystatic` segment should become a Keystatic Admin UI route.

You can leverage Astro's [Rest parameters](/en/core-concepts/routing/#rest-parameters) to match file paths of any depth.

Create a new page called `src/pages/keystatic/[...params].astro`:

```ts
---
// src/pages/keystatic/[...params].astro

// TODO: Display Keystatic pages
---
```

The Keystatic Admin UI is built with React. So that you can import and mount this app here with React, create a new file outside of the pages directory.

You can put it in the project root, alongside your Keystatic config file.

```ts
// keystatic.page.ts
import { makePage } from '@keystatic/astro/ui'
import keystaticConfig from './keystatic.config'

export const Keystatic = makePage(keystaticConfig)
```

> Now, you can import that file in our `keystatic/[...params].astro` page, and mount the component on the clientside only:

Reviewer 3: ```suggestion
Now, you can import that file in our `keystatic/[...params].astro` page, and mount the component on the client side only:
```

```ts
---
// src/pages/keystatic/[...params].astro
import { Keystatic } from '../../../keystatic.page'
---

<Keystatic client:only />
```

Next, start up Astro's integrated server:

```sh
npm run dev
```

### Visiting the Keystatic Admin UI

At this point, if you try to visit the `/keystatic` route in the browser, you'll get an error:

<Aside title="error" type="danger">
  `getStaticPaths()` function is required for dynamic routes. Make sure that you export a `getStaticPaths` function from your dynamic route.
  
  Alternatively, set `output: "server"` in your Astro config file to switch to a non-static server build.
</Aside>

You *could* do this, but instead, use Astro's new [`hybrid` rendering mode](/en/guides/server-side-rendering/#converting-a-static-site-to-hybrid-rendering), which works even better here.

### Astro's SSR hybrid mode

In the `astro.config.mjs`, add the `output: 'hybrid'` option:

```diff title="astro.config.mjs"
export default defineConfig({

> + output: 'hybrid',

Reviewer 4: ```suggestion
+  output: 'hybrid',
```

  integrations: [react()],
})
```

Hybrid rendering will still try to pre-render every page by default, with a mechanism to opt out of it.

You *don't want* prerendering on Keystatic routes. Export a `const prerender = false` in your Keystatic pages file:

```diff
---
// src/pages/keystatic/[...params].astro
import { Keystatic } from '../../../keystatic.page'

> + export const prerender = false

Reviewer 4: ```suggestion
+export const prerender = false
```

})

export const prerender = false
```

Notice those routes also opt out of prerendering.

Aaaand... You should be all set to use Keystatic now 🎉

:::note
At this point, you should be able to visit the `/keystatic` route in the browser and see the Keystatic Admin UI. 
:::

## Creating a new post

Try and visit the `/keystatic` page in the browser one more time, and click on the “Posts” collection.

Go ahead, and create a new post.

:::note
In our Keystatic config file, you've set the `path` property for our `posts` collection to `src/content/posts/*`.

As a result, creating a new post from the Keystatic Admin UI should create a new `content` directory in the `src` directory, with the new post `.mdoc` file inside.
:::

If everything worked correctly, you'll find your new post inside the `src/content/posts` directory:

<FileTree title="Project Structure">
- src/
  - content/
    - posts/
      - **my-first-post.mdoc**
</FileTree>

That Markdoc file should look something like this:

```markdown
---
title: My First Post
---

This is my very first post. I am **super** excited.
```

Niiiice ✨

Time to display that post on the frontend now.

## Rendering Keystatic content

:::tip

> While Keystatic provides its own [Reader and Renderer APIs](https://keystatic.com/docs/reader-api) to pull data your frontend, it also work great with Astro's [content collections](/en/guides/content-collections/)!

Reviewer 3: ```suggestion
While Keystatic provides its own [Reader and Renderer APIs](https://keystatic.com/docs/reader-api) to pull data for your frontend, it also works great with Astro's [content collections](/en/guides/content-collections/)!
```

:::

### Displaying a collection list

> Since you told Keystatic to manage `posts` in `src/content/posts`, they'll be available as Astro's content collections as the `posts` collection slug:

Reviewer 3: ```suggestion
Since you instructed Keystatic to manage `posts` in `src/content/posts`, they'll be available as Astro's content collections as the `posts` collection slug:
```

```tsx {4}
---
import { getCollection } from 'astro:content'

const posts = await getCollection('posts')
---

<main>
  <pre>{JSON.stringify(posts, null, 2)}</pre>
</main>

```

### Displaying a single collection entry

You can also get a single collection entry by its slug:

```tsx {4}
---
import { getEntry } from 'astro:content'

const post = await getEntry('posts', 'my-first-post')
---

<main>
  <pre>{JSON.stringify(post, null, 2)}</pre>
</main>
```

### Rendering document field content

> If you look at the output from our `post` entry from above, you'll notice that the `body` field – the `content` of our post – is displaying raw Markdoc.

Reviewer 3: ```suggestion
If you examine the output from our `post` entry above, you will notice that the `body` field — which represents the `content` of our post — is displaying raw Markdown.
```
